### DISCUSSION TITLE
[Messenger] Is there a way to auto-stop a worker after a certain idle duration?

### DISCUSSION THREAD

**Discussion by maidmaid** [symfony/symfony#54185] [Q&A] [answered]
I run the messenger worker in [Google Cloud Run Jobs](https://cloud.google.com/run/docs/create-jobs) which is a serverless environnement. To reduce usages (and costs!), I'm thinking about to _scale to 0_ the worker when idle.

The solution could be to schedule the worker every 1 hour and make it able to stop itself if it's idle for 30 seconds. Thus, if you have a few messages to process, the worker would work 24 * some minutes a day instead to have to work full time a day (24 * 1h).

**Comment by maidmaid** [comment]
friendly ping @lyrixx, do you have an idea?

**Reply by lyrixx** [reply]
I just blogged about this (but in french, sorry): https://jolicode.com/blog/forcer-le-redemarrage-dun-worker-symfony . You can create your own logic in the listener then, stop the worker like in the code

Or you could also use the `--time-limit` option of `messenger:consume` command

```
  Use the --time-limit option to stop the worker when the given time limit (in seconds) is reached.
  If a message is being handled, the worker will stop after the processing is finished:
  
      php bin/console messenger:consume <receiver-name> --time-limit=3600
```

**Reply by maidmaid** [reply]
Thanks for your answer ! `--time-limit=3600` stops the worker after 1h, regardless if it's idle. But I can maybe create a listener which _watches_ the idle time and once the limit reached, I throw a `StopWorkerExceptionInterface` to stop the worker, as suggested in your article. I don't know how possible it is.

**Reply by lyrixx** [reply, marked as answer]
You save the date of the last message (by listeneing on message handled or message failure), and worker idle you take a decision 

Take [this listener](https://github.com/symfony/symfony/blob/7.1/src/Symfony/Component/Messenger/EventListener/StopWorkerOnFailureLimitListener.php) as example

**Reply by maidmaid** [reply]
I think we can't wait a idle moment before stopping. But stopping immediately after the idle status seems possible. It'll do it. Thank you for your help !
